chrometool/chrometool_test.go

```diff
@@ -0,0 +1,317 @@
+				serverText := strings.Replace(scanner.Text(), responseID, requestID, 1)
+				serverData = []byte(serverText)
+			}
+
+			// The delay here is intentional to avoid race conditions that chromedp sometimes encounters
+			// removing this may lead test to hang and timeout
+			// TODO: ensure prevention of such race conditions at upstream
+			time.Sleep(1 * time.Millisecond)
+			wsutil.WriteMessage(conn, ws.StateServerSide, op, serverData)
+		}
+	}
+}
+
+func populateResponses(log string) (map[requestType]responseType, error) {
+	responseMap := make(map[requestType]responseType)
+
+	logFile, err := os.Open(log)
+	if err != nil {
+		return nil, err
+	}
+
+	defer func() {
+		logFile.Close()
+	}()
+
+	scanner := bufio.NewScanner(logFile)
+	var serverData []byte
+	var prevRequestType requestType
+	IDmap := make(map[interface{}]requestType)
+
+	//populate prevRequestType
+	for scanner.Scan() {
+		var entry map[string]interface{}
+		json.Unmarshal(scanner.Bytes(), &entry)
+		if id, ok := entry["id"]; ok {
+			IDmap[id] = newRequestType(entry)
+			prevRequestType = newRequestType(entry)
+			break
+		}
+	}
+
+	for scanner.Scan() {
+		var entry map[string]interface{}
+
+		json.Unmarshal(scanner.Bytes(), &entry)
+		if id, ok := entry["id"]; ok {
+			if reqType, isPresent := IDmap[id]; isPresent {
+				line := append(scanner.Bytes(), '\n')
+				serverData = append(serverData, line...)
+				if resType, isPres := responseMap[reqType]; isPres {
+					resType.responseList = append(resType.responseList, serverData)
+					responseMap[reqType] = resType
+				} else {
+					responseMap[reqType] = newResponseType(serverData)
+				}
+				prevRequestType = reqType
+				delete(IDmap, id)
+			} else {
+				resType := responseMap[prevRequestType]
+				resList := resType.responseList
+				resList[len(resList)-1] = append(resList[len(resList)-1], serverData...)
+				resType.responseList = resList
+				responseMap[prevRequestType] = resType
+				IDmap[id] = newRequestType(entry)
+			}
+			serverData = nil
+			continue
+		}
+		line := append(scanner.Bytes(), '\n')
+		serverData = append(serverData, line...)
+	}
+	err = scanner.Err()
+	return responseMap, err
+}
+
+func newResponseType(response []byte) responseType {
+	var responses [][]byte
+	res := responseType{
+		responseList: append(responses, response),
+	}
+	return res
+}
+
+func newRequestType(request map[string]interface{}) requestType {
+	req := requestType{}
+	if method, ok := request["method"]; ok {
+		reqMethod, err := json.Marshal(method)
+		if err != nil {
+			fmt.Println(err)
+		}
+		req.method = string(reqMethod)
+	}
+	if params, ok := request["params"]; ok {
+		reqParams, err := json.Marshal(params)
+		if err != nil {
+			fmt.Println(err)
+		}
+		req.params = string(reqParams)
+	}
+	return req
+}
+
+// testServer is a simple HTTP server that displays the passed headers in the html.
+func testServer(t *testing.T, addr string, run bool) error {
+	if !run {
+		return nil
+	}
```

Comment:
why take a bool only to short circuit the function?
Remove the bool, and call the function only if run is true?